nuxeo-core/nuxeo-core-storage-mongodb/src/main/java/org/nuxeo/ecm/core/storage/mongodb/MongoDBRepository.java

```diff
@@ -429,15 +424,63 @@ protected void update(String op, String key, Object value) {
          * <p>
          * A conflict occurs if one key is equals to or is a prefix of the other.
          */
-        protected boolean conflicts(String key, List<String> previousKeys) {
-            String keydot = key + '.';
-            for (String prev : previousKeys) {
-                if (prev.equals(key) || prev.startsWith(keydot) || key.startsWith(prev + '.')) {
+        protected void checkForConflict(String key) {
+            List<String> pKeys = getPrefixKeys(key);
+            if (conflictKeys(key, pKeys)) {
+                newUpdate();
+            }
+            prefixKeys.addAll(pKeys);
+            keys.add(key);
+        }
+
+        protected void newUpdate() {
+            updateList.add(update = new BasicDBObject());
+            int pSize = 10;
+            int size = 10;
+            if (prefixKeys != null) {
+                pSize = prefixKeys.size();
+            }
+            if (keys != null) {
+                size = keys.size();
+            }
+            prefixKeys = new HashSet<>(pSize);
+            keys = new HashSet<>(size);
+        }
+
+        protected boolean conflictKeys(String key, List<String> subkeys) {
+            if (prefixKeys.contains(key)) {
+                return true;
+            }
+            for (String sk: subkeys) {
+                if (keys.contains(sk)) {
                     return true;
                 }
             }
             return false;
         }
+
+        /**
+         * return a list of parents key
+         * foo.0.bar -> [foo, foo.0, foo.0.bar]
+         */
+        protected List<String> getPrefixKeys(String key) {
+            List<String> ret = new ArrayList<>(10);
+            // much faster than regex String.split
+            Iterable<String> split = Splitter.on('.').split(key);
+            String path = null;
+            for (String part : split) {
+                if (path == null) {
+                    path = part;
+                } else {
+                    path += "." + part;
+                }
+                ret.add(path);
+            }
+            if (path == null) {
+                ret.add(key);
+            }
+            return ret;
```

Comment:
I think just iterating of the string to find `.` characters and creating a new substring each time would be even faster, as there would be just substring allocations which are very fast (reuse the `char[]` of the string without new allocation).